--- uim/xim/convdisp.cpp	2019-06-19 21:33:44.829429500 +0900
+++ uim.new/xim/convdisp.cpp	2019-06-19 21:40:59.517137300 +0900
@@ -54,7 +54,8 @@
 #include "uim/uim-scm.h"
 
 #define UNDERLINE_HEIGHT	2
-#define DEFAULT_FONT_SIZE	16
+#define SCALE_FACTOR		2
+#define DEFAULT_FONT_SIZE	(16 * SCALE_FACTOR)
 // Temporal hack for flashplayer plugin's broken over-the-spot XIM style
 #define FLASHPLAYER7_WORKAROUND
 #define FLASHPLAYER9_WORKAROUND
@@ -452,7 +453,7 @@
 void PeWin::expose(Window w)
 {
     XCopyArea(XimServer::gDpy, mPixmap, w, mGC,
-	      0, 0, mWidth, mHeight, 0, 0);
+	      0, 0, mWidth * SCALE_FACTOR, mHeight * SCALE_FACTOR, 0, 0);
 
     XFlush(XimServer::gDpy);
 }
@@ -598,9 +599,9 @@
     if (w == mWidth && h == mHeight)
 	return;
 
-    XResizeWindow(XimServer::gDpy, mWin, w, h);
+    XResizeWindow(XimServer::gDpy, mWin, w * SCALE_FACTOR, h * SCALE_FACTOR);
     XFreePixmap(XimServer::gDpy, mPixmap);
-    mPixmap = XCreatePixmap(XimServer::gDpy, DefaultRootWindow(XimServer::gDpy), w, h,
+    mPixmap = XCreatePixmap(XimServer::gDpy, DefaultRootWindow(XimServer::gDpy), w * SCALE_FACTOR, h * SCALE_FACTOR,
 			    DefaultDepth(XimServer::gDpy, DefaultScreen(XimServer::gDpy)));
 #if HAVE_XFT_UTF8_STRING 
     if (mConvdisp->use_xft() == true)
@@ -631,7 +632,7 @@
 void PeWin::clear()
 {
     XFillRectangle(XimServer::gDpy, mPixmap, mClearGC,
-		   0, 0, mWidth, mHeight);
+		   0, 0, mWidth * SCALE_FACTOR, mHeight * SCALE_FACTOR);
 }
 
 void PeWin::draw()
@@ -654,10 +655,10 @@
 //
 // PeLineWin
 //
-#define PE_LINE_WIN_WIDTH	400
-#define PE_LINE_WIN_HEIGHT	28
-#define PE_LINE_WIN_FONT_POS_Y	20
-#define PE_LINE_WIN_MARGIN_X	2
+#define PE_LINE_WIN_WIDTH	(400 * SCALE_FACTOR)
+#define PE_LINE_WIN_HEIGHT	( 28 * SCALE_FACTOR)
+#define PE_LINE_WIN_FONT_POS_Y	( 20 * SCALE_FACTOR)
+#define PE_LINE_WIN_MARGIN_X	(  2 * SCALE_FACTOR)
 
 PeLineWin::PeLineWin(Window w, const char *im_lang, const char *encoding, const char *locale, Convdisp *cd) : PeWin(w, im_lang, encoding, locale, cd)
 {
@@ -817,14 +818,14 @@
     if (m_mask_pix) {
 	XFreePixmap(XimServer::gDpy, m_mask_pix);
 	m_mask_pix = XCreatePixmap(XimServer::gDpy, mWin,
-				   mWidth, mHeight, 1);
+				   mWidth * SCALE_FACTOR, mHeight * SCALE_FACTOR, 1);
     }
 }
 
 void PeOvWin::draw_ce(char_ent *ce, int len)
 {
     if (m_mask_pix == 0) {
-	m_mask_pix = XCreatePixmap(XimServer::gDpy, mWin, mWidth, mHeight, 1);
+	m_mask_pix = XCreatePixmap(XimServer::gDpy, mWin, mWidth * SCALE_FACTOR, mHeight * SCALE_FACTOR, 1);
 	m_mask_pix_gc = XCreateGC(XimServer::gDpy, m_mask_pix, 0, 0);
     }
 
@@ -834,7 +835,7 @@
 		   BlackPixel(XimServer::gDpy,
 			      DefaultScreen(XimServer::gDpy)));
     XFillRectangle(XimServer::gDpy, m_mask_pix,
-		   m_mask_pix_gc, 0, 0, mWidth, mHeight);
+		   m_mask_pix_gc, 0, 0, mWidth * SCALE_FACTOR, mHeight * SCALE_FACTOR);
     XSetForeground(XimServer::gDpy, m_mask_pix_gc,
 		   WhitePixel(XimServer::gDpy,
 			      DefaultScreen(XimServer::gDpy)));
